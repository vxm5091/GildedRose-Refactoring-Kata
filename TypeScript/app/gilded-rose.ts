import { ICustomItem } from './interfaces';
import {
  AgedBrie,
  BackstagePasses,
  ConjuredItem,
  Item,
  LegendaryItem,
  RegularItem,
} from './inventory';

export class GildedRose {
  
  // stores the respective custom class for each esoteric item
  // removes hard coded logic from getCustomItem()
  private static itemClassReference = [
    {
      name: 'Backstage passes to a TAFKAL80ETC concert',
      class: BackstagePasses,
    },
    { name: 'Aged Brie', class: AgedBrie },
    { name: 'Sulfuras, Hand of Ragnaros', class: LegendaryItem },
    { pattern: 'Conjured', class: ConjuredItem },
  ];
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  // find matching class based on full name match or partial pattern match ('Conjured')
  // otherwise return RegularItem instance
  getCustomItem(item: Item): ICustomItem {
    const itemClass =
      GildedRose.itemClassReference.find(
        ic =>
          item.name === ic.name ||
          (ic.pattern && item.name.includes(ic.pattern)),
      )?.class || RegularItem;
    return new itemClass(item);
  }

  updateQuality() {
    return this.items.map(item => {
      const customItem = this.getCustomItem(item);
      customItem.update()
      return customItem.item;
    });
  }
}
