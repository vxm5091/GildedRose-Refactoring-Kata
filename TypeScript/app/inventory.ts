import { ICustomItem } from './interfaces';

export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name: string, sellIn: number, quality: number) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

/* ------------------------------------------ Base Class ------------------------------------------ */
/**
 * abstracts the updating of quality and sellIn values
 *  */
class CustomItemBase {
  item: Item;
  private readonly _maxQuality = 50;
  private readonly _minQuality = 0;

  constructor(item: Item) {
    this.item = item;
  }

  private _updateQuality(change: number, isLegendary = false) {
    let updatedQuality = this.item.quality + change;

    // logic doesn't apply to legendary item which remains at constant quality or possibly increases
    if (!isLegendary) {
      updatedQuality = Math.min(updatedQuality, this._maxQuality); // can't go above max value (50)
      updatedQuality = Math.max(updatedQuality, this._minQuality); // can't go below min value (0)
    }

    this.item.quality = updatedQuality;
  }

  private _updateSellIn(isLegendary = false) {
    if (isLegendary) return;
    this.item.sellIn--;
  }

  public update(change: number, isLegendary = false) {
    this._updateSellIn(isLegendary);
    this._updateQuality(change, isLegendary);
  }
}

/* ------------------------------------------ Custom Item Classes ------------------------------------------ */
/**
 * @method getRateOfQualityChange() defines item-specific logic
 * @property isLegendary (optional) bypasses max quality value and sellIn change
 *  */
export class RegularItem extends CustomItemBase implements ICustomItem {
  getRateOfQualityChange() {
    let change = -1;
    if (this.item.sellIn < 0) change = -2;
    return change;
  }
}

export class BackstagePasses extends CustomItemBase implements ICustomItem {
  getRateOfQualityChange() {
    const { sellIn, quality } = this.item;
    if (sellIn < 0) return -quality; // worthless after concert date
    let change = 1;
    if (sellIn < 11) change++; // Quality increases by 1 extra point if <11 days to concert (2 total)
    if (sellIn < 6) change++; // Quality increases by 1 extra point if <6 days to concert (3 total)
    return change;
  }
}

export class LegendaryItem extends CustomItemBase implements ICustomItem {
  isLegendary = true;
  getRateOfQualityChange() {
    return 0;
  }
}

export class AgedBrie extends CustomItemBase implements ICustomItem {
  getRateOfQualityChange() {
    if (this.item.quality < 50) return 1;
    return 0;
  }
}

export class ConjuredItem extends CustomItemBase implements ICustomItem {
  getRateOfQualityChange() {
    // 	degrades in quality 2x rate of normal item
    // 	get rate of change if this were a normal item
    const regularItem = new RegularItem(this.item);
    const regularChange = regularItem.getRateOfQualityChange();
    return regularChange * 2;
  }
}
