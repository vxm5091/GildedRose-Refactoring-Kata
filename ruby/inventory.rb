class Item
  attr_accessor :name, :sell_in, :quality

  def initialize(name, sell_in, quality)
    @name = name
    @sell_in = sell_in
    @quality = quality
  end

  def to_s()
    "#{@name}, #{@sell_in}, #{@quality}"
  end
end

# ------------------------------------------ Base Class ------------------------------------------ */
# abstracts the updating of quality and sellIn values in the Item class dependency
class CustomItemBase
  attr_accessor :item

  def initialize(item)
    @item = item
    @max_quality = 50
    @min_quality = 0
  end

  def legendary?
    false
  end

  private

  attr_reader :max_quality, :min_quality

  def update_quality(change)
    updated_quality = @item.quality + change
    unless legendary?
      updated_quality = [updated_quality, max_quality].min
      updated_quality = [updated_quality, min_quality].max
    end

    @item.quality = updated_quality
  end

  def update_sell_in
    @item.sell_in -= 1 unless legendary?
  end

  public

  def update(change)
    update_sell_in
    update_quality(change)
  end
end


# ------------------------------------------ Custom Item Classes ------------------------------------------ */
# getRateOfQualityChange() defines item-specific logic
# isLegendary (optional) bypasses max quality value and sellIn change
class RegularItem < CustomItemBase
  def rate_of_quality_change
    change = -1
    change = -2 if item.sell_in.negative?
    change
  end
end

class BackstagePasses < CustomItemBase
  def rate_of_quality_change
    sell_in = item.sell_in
    quality = item.quality

    return -quality if sell_in.negative?

    change = 1
    change += 1 if sell_in < 11
    change += 1 if sell_in < 6

    change
  end
end

class LegendaryItem < CustomItemBase
  def rate_of_quality_change
    0
  end

  def legendary?
    true
  end
end

class AgedBrie < CustomItemBase
  def rate_of_quality_change
    1
  end
end

class ConjuredItem < CustomItemBase
  def rate_of_quality_change
    # 	degrades in quality 2x rate of normal item
    # 	get rate of change if this were a normal item
    regular_item = RegularItem.new(item)
    regular_change = regular_item.rate_of_quality_change
    regular_change * 2
  end
end


