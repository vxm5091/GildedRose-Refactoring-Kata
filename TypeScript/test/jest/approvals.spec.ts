import { GildedRose } from '@/gilded-rose';
import { Item } from '@/inventory';

describe('Gilded Rose Approval', () => {
  let gameConsoleOutput: string;
  let originalConsoleLog: (message: any) => void;
  let originalProcessArgv: string[];

  function gameConsoleLog(msg: string) {
    if (msg) {
      gameConsoleOutput += msg;
    }
    gameConsoleOutput += '\n';
  }

  beforeEach(() => {
    // prepare capturing console.log to our own gameConsoleLog.
    gameConsoleOutput = '';
    originalConsoleLog = console.log;
    console.log = gameConsoleLog;
    originalProcessArgv = process.argv;
  });

  afterEach(() => {
    // reset original console.log
    console.log = originalConsoleLog;
    process.argv = originalProcessArgv;
  });

  it('should foo', () => {
    // const gildedRose = new GildedRose([new Item('Backstage passes to a TAFKAL80ETC concert', 15, 20)]);
    const items = [
      new Item('+5 Dexterity Vest', 10, 20), //
      new Item('Aged Brie', 2, 0), //
      new Item('Elixir of the Mongoose', 5, 7), //
      new Item('Sulfuras, Hand of Ragnaros', 0, 80), //
      new Item('Sulfuras, Hand of Ragnaros', -1, 80),
      new Item('Backstage passes to a TAFKAL80ETC concert', 15, 20),
      new Item('Backstage passes to a TAFKAL80ETC concert', 10, 49),
      new Item('Backstage passes to a TAFKAL80ETC concert', 5, 49),
      // this conjured item does not work properly yet
      new Item('Conjured Mana Cake', 3, 6),
    ];

    const gildedRose = new GildedRose(items);
    const updatedItems = gildedRose.updateQuality();

    expect(updatedItems).toMatchSnapshot();
  });
});
