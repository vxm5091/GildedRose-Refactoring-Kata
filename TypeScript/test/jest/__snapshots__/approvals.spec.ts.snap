// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gilded Rose Approval should foo 1`] = `
[
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 19,
    "sellIn": 9,
  },
  Item {
    "name": "Aged Brie",
    "quality": 1,
    "sellIn": 1,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 6,
    "sellIn": 4,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 21,
    "sellIn": 14,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 9,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 4,
  },
  Item {
    "name": "Conjured Mana Cake",
    "quality": 4,
    "sellIn": 2,
  },
]
`;
